File: Cmdlets/GetExchangeRateCommand.vb
Imports System
Imports System.Collections.Generic
Imports System.Management.Automation
Imports TIKSN.Finance

<Cmdlet("Get", "ExchangeRate")>
Public Class GetExchangeRateCommand
	Inherits Cmdlet

	<Parameter(Mandatory:=True)>
	Public Property BaseCurrency As String

	<Parameter(Mandatory:=True)>
	Public Property CounterCurrency As String

	<Parameter(Mandatory:=False)>
	Public Property AsOn As DateTimeOffset?

	Protected Overrides Sub BeginProcessing()
		Dim banks = GetSupportedBanks()
		Dim exchangeDate = If(AsOn.HasValue, AsOn.Value, DateTimeOffset.Now)

		Dim result As New List(Of BankExchangeRate)

		Dim pair As New CurrencyPair(New CurrencyInfo(BaseCurrency), New CurrencyInfo(CounterCurrency))
		Dim BankProgressRecord As New ProgressRecord(0, "Retrieving exchange rates from bank", "Bank name")

		For Each bank In banks

			BankProgressRecord.StatusDescription = bank.BankName
			WriteProgress(BankProgressRecord)

			WriteVerbose(String.Format("Retrieving supported currencies from {0}", bank.BankName))

			Try
				Dim pairs = bank.Exchange.GetCurrencyPairsAsync(exchangeDate).Result

				If pairs.Any(Function(item) item.Equals(pair)) Then
					WriteVerbose(String.Format("Retrieving {1} exchange rate from {0}", bank.BankName, pair))
					Dim rate = bank.Exchange.GetExchangeRateAsync(pair, exchangeDate).Result
					result.Add(New BankExchangeRate(bank.BankName, pair, rate))
				End If
			Catch ex As Exception
				WriteError(New ErrorRecord(ex, "", ErrorCategory.InvalidResult, Nothing))
			End Try
		Next

		WriteObject(result, True)

		If result.Count = 0 Then
			WriteWarning(String.Format("Exchange rate for {0} is not found.", pair))
		End If

		BankProgressRecord.RecordType = ProgressRecordType.Completed
		WriteProgress(BankProgressRecord)
	End Sub
End Class

File: Cmdlets.Core/Commands/GetExchangeRateCommand.cs
using Microsoft.Extensions.DependencyInjection;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Management.Automation;
using System.Threading;
using System.Threading.Tasks;
using TIKSN.Cmdlets.Commands;
using TIKSN.Cmdlets.Common.Helpers;
using TIKSN.Cmdlets.Common.Models;
using TIKSN.Finance;
using TIKSN.Time;

namespace TIKSN.Cmdlets.Core.Commands
{
    [Cmdlet("Get", "ExchangeRate")]
    public class GetExchangeRateCommand : Command
    {
        [Parameter(Mandatory = false)]
        public DateTimeOffset? AsOn { get; set; }

        [Parameter(Mandatory = true)]
        public string BaseCurrency { get; set; }

        [Parameter(Mandatory = true)]
        public string CounterCurrency { get; set; }

        protected override async Task ProcessRecordAsync(CancellationToken cancellationToken)
        {
            var banks = ExchangeHelper.GetSupportedBanks();
            var timeProvider = ServiceProvider.GetRequiredService<ITimeProvider>();
            var exchangeDate = AsOn.HasValue ? AsOn.Value : timeProvider.GetCurrentTime();

            List<BankExchangeRate> result = new List<BankExchangeRate>();

            CurrencyPair pair = new CurrencyPair(new CurrencyInfo(BaseCurrency), new CurrencyInfo(CounterCurrency));
            ProgressRecord BankProgressRecord = new ProgressRecord(0, "Retrieving exchange rates from bank", "Bank name");

            foreach (var bank in banks)
            {
                BankProgressRecord.StatusDescription = bank.BankName;
                WriteProgress(BankProgressRecord);

                WriteVerbose(string.Format("Retrieving supported currencies from {0}", bank.BankName));

                try
                {
                    var pairs = await bank.Exchange.GetCurrencyPairsAsync(exchangeDate, cancellationToken);

                    if (pairs.Any(item => item.Equals(pair)))
                    {
                        WriteVerbose(string.Format("Retrieving {1} exchange rate from {0}", bank.BankName, pair));
                        var rate = await bank.Exchange.GetExchangeRateAsync(pair, exchangeDate, cancellationToken);
                        result.Add(new BankExchangeRate(bank.BankName, pair, rate));
                    }
                }
                catch (Exception ex)
                {
                    WriteError(new ErrorRecord(ex, "", ErrorCategory.InvalidResult, null));
                }
            }

            WriteObject(result, true);

            if (result.Count == 0)
            {
                WriteWarning(string.Format("Exchange rate for {0} is not found.", pair));
            }

            BankProgressRecord.RecordType = ProgressRecordType.Completed;
            WriteProgress(BankProgressRecord);
        }
    }
}